libs/node-custom/src/usb_device.ts

```diff
@@ -0,0 +1,265 @@
+        );
+        await this.device.clearHalt('in', readEndpointNumber);
+        continue;
+      } else if (!inResult.data || inResult.data.byteLength === 0) {
+        debug('received no data, aborting');
+        break;
+      }
+
+      const data = Buffer.from(inResult.data.buffer);
+      debug(
+        'received response: %o (%s)',
+        data,
+        inspect(truncateStringForDisplay(data.toString('utf8')))
+      );
+      return ok(data);
+    }
+
+    debug('retries exhausted, aborting');
+    return err(ErrorCode.NoDeviceAnswer);
+  }
+
+  /**
+   * Low-level method to send data to the device.
+   */
+  async write(data: Buffer): Promise<Result<void, ErrorCode>> {
+    if (!this.connected) {
+      return err(ErrorCode.ScannerOffline);
+    }
+
+    const getEndpointsResult = this.getReadWriteEndpoints();
+
+    if (getEndpointsResult.isErr()) {
+      return getEndpointsResult;
+    }
+
+    let outResult!: USBOutTransferResult;
+    const [, writeEndpointNumber] = getEndpointsResult.ok();
+
+    for (let i = 0; i < WRITE_RETRY_MAX; i += 1) {
+      debug('sending request (attempt %d/%d)', i + 1, WRITE_RETRY_MAX);
+      outResult = await this.device.transferOut(writeEndpointNumber, data);
+
+      if (outResult.status !== 'ok') {
+        debug('send failed (status=%s)', outResult.status);
+        await this.device.clearHalt('out', writeEndpointNumber);
+        break;
+      }
+
+      if (outResult.bytesWritten !== data.length) {
+        debug(
+          'send failed (bytesWritten=%d, expected=%d)',
+          outResult.bytesWritten,
+          data.length
+        );
+        continue;
+      }
+
+      debug(
+        'sent request: %o (%s)',
+        data,
+        inspect(truncateStringForDisplay(data.toString('utf8')))
+      );
+      return ok();
+    }
+
+    return err(ErrorCode.WriteError);
+  }
+
+  /**
+   * Performs a `write` followed by a `read`.
+   */
+  async writeRead(
+    data: Buffer,
+    maxLength: number
+  ): Promise<Result<Buffer, ErrorCode>> {
+    const writeResult = await this.write(data);
+
+    if (writeResult.isErr()) {
+      return writeResult;
+    }
+
+    return this.read(maxLength);
+  }
+
+  /**
+   * Runs the given function with the device locked.
+   */
+  async withLock<T>(fn: () => Promise<T>): Promise<T> {
+    await this.lock.acquire();
+
+    try {
+      return await fn();
+    } finally {
+      this.lock.release();
+    }
+  }
+
+  /**
+   * Gets the read and write endpoints for communicating with the device.
+   */
+  private getReadWriteEndpoints(): Result<
+    [read: number, write: number],
+    ErrorCode
+  > {
+    try {
+      const readEndpointNumber =
+        this.device.configurations[0]?.interfaces[INTERFACE_INDEX]?.alternate
+          .endpoints[READ_ENDPOINT_INDEX]?.endpointNumber;
+      const writeEndpointNumber =
+        this.device.configurations[0]?.interfaces[INTERFACE_INDEX]?.alternate
+          .endpoints[WRITE_ENDPOINT_INDEX]?.endpointNumber;
```

Comment:
Are these endpoint numbers not constant? Why not hardcode them?

Reply:
I've generally tried to give names to the magic numbers in our codebase so we know what they're for.